python/understack-workflows/understack_workflows/main/enroll_server.py

```diff
@@ -0,0 +1,140 @@
+    verbatim to the newly created server Device).
+
+    The server Device type must exist, with a name that matches the "model" as
+    reported by the BMC.
+
+    The DRAC IP Prefix must exist.
+
+    This script has the following order of operations:
+
+    - connect to the BMC, trying standard password then factory default
+
+    - ensure standard BMC password is set
+
+    - if DHCP, set permanent IP address, netmask, default gw
+
+    - TODO: create and install SSL certificate
+
+    - TODO: update BMC firmware
+
+    - TODO: set NTP Server IPs for DRAC
+      (NTP server IP addresses are different per region)
+
+    -  Using BMC, configure our standard BIOS settings
+       - set PXE boot device
+       - set timezone to UTC
+
+    -  from BMC, discover basic hardware info:
+       - manufacturer, model number, serial number
+       - list ethernet interfaces with:
+          - name like BMC or SLOT.NIC.1-1
+          - MAC address
+          - LLDP connections [{remote_mac, remote_interface_name}]
+
+    - Find or create this server in Nautobot by serial number.
+
+    - set name, manufacturer, model, serial, location, rack
+
+    - Find BMC interface
+
+    - For each server interface
+        - find or create server interface by name in nautobot
+        - set interface mac addresses
+        - look up switch by mac addr (is stored in Nautobot's asset tag field)
+        - look up switch interface by name
+        - find or create cable
+
+    - create BMC IP address assignment for BMC interface - convert our type
+      "dhcp" IP Address to type "host" and associate it with the interface
+
+    -  Find or create this baremetal node in Ironic
+       - TODO create ports with MACs (omit BMC port) and set one to PXE
+       - TODO advance to available state
+       - TODO set flavor?  what else?
+
+    """
+    args = argument_parser().parse_args()
+
+    bmc_ip_address = args.bmc_ip_address
+    logger.info(f"{__file__} starting for {bmc_ip_address=}")
+
+    url = args.nautobot_url
+    token = args.nautobot_token or credential("nb-token", "token")
+    nautobot = pynautobot.api(url, token=token)
+
+    bmc = bmc_for_ip_address(bmc_ip_address, password=args.bmc_password)
+    logger.info("Checking BMC credentials")
+    set_bmc_password(bmc.ip_address, bmc.password)
+
+    device_info = chassis_info(bmc)
+
+    # TODO: make this pseudo-idempotent by ignoring the error when a job is
+    # already scheduled:
+    update_dell_drac_settings(bmc)
+    update_dell_bios_settings(bmc)
+
+    logger.info(f"Discovered {device_info}")
+    device = nautobot_device.find_or_create(device_info, nautobot)
+
+    # Do this after Nautobot IPAddress has been changed from DHCP!
+    bmc_set_permanent_ip_addr(bmc, device_info.bmc_interface)
+
+    _ironic_provision_state = ironic_node.create_or_update(
+        device["id"], device["name"], device_info.manufacturer, bmc, logger
+    )
+    logger.info(f"{device['id']} {_ironic_provision_state=}")
```

Comment:
So realistically this is what the workflow steps would be. There's already a few entry points matching this that exist in the existing code base that should be updated to this. Right now this whole PR duplicates a bunch of code without cleaning anything up or making the pieces reusable by having alternate entry points